crates/ruff_linter/src/rules/flake8_unused_arguments/rules/unused_arguments.rs

```diff
@@ -311,6 +311,71 @@ fn call<'a>(
     }));
 }
 
+/// Returns `true` if a function appears to be a base class stub. In other
+/// words, if it matches the following syntax:
+///
+/// ```text
+/// variable = <string | f-string>
+/// raise NotImplementedError(variable)
+/// ```
+///
+/// See also [`is_stub`]. We're a bit more generous in what is considered a
+/// stub in this rule to avoid clashing with [`EM101`].
+///
+/// [`is_stub`]: function_type::is_stub
+/// [`EM101`]: https://docs.astral.sh/ruff/rules/raw-string-in-exception/
+fn is_not_implemented_stub_with_variable(function_def: &StmtFunctionDef) -> bool {
+    // Ignore doc-strings.
+    let statements = match function_def.body.as_slice() {
+        [Stmt::Expr(StmtExpr { value, .. }), rest @ ..]
+            if matches!(value.as_ref(), ast::Expr::StringLiteral(_)) =>
+        {
+            rest
+        }
+        _ => &function_def.body,
+    };
+
+    let [first_stmt, second_stmt] = statements else {
+        return false;
+    };
+
+    let Stmt::Assign(ast::StmtAssign { value, .. }) = first_stmt else {
+        return false;
+    };
+
+    if !matches!(
+        value.as_ref(),
+        ast::Expr::StringLiteral(_) | ast::Expr::FString(_)
+    ) {
+        return false;
+    }
+
+    let Stmt::Raise(StmtRaise {
+        exc: Some(exception),
+        ..
+    }) = second_stmt
+    else {
+        return false;
+    };
+
+    let ast::Expr::Call(ast::ExprCall {
+        func, arguments, ..
+    }) = exception.as_ref()
```

Comment:
Nit:
```suggestion
    }) = &**exception
```

Reply:
I see mixed usages of `&**` and `as_ref` throughout the code base. Out of curiosity, when is one used over the other? The general heuristic I've seen is `*` and `&` are recommended for simple stuff; and, once you start de-referencing and re-borrowing, that `as_ref` is recommended if just to avoid needing to think about whatever container the variable is in.

Reply:
We used to use `as_ref` but started to prefer dereferencing (except for `Option` where you still need `as_ref` and `as_deref`) because it's possible that `as_ref` becomes ambiguous if a new `AsRef` implementation is added to a type. For example, `String` has many `as_ref` methods:

```
  = note: multiple `impl`s satisfying `String: AsRef<_>` found in the following crates: `alloc`, `std`:
          - impl AsRef<OsStr> for String;
          - impl AsRef<Path> for String;
          - impl AsRef<[u8]> for String;
          - impl AsRef<str> for String;
```

Reply:
Oh interesting, that makes sense! Thank you!